File: C++/pancake_sort.cpp
#include <iostream>

using namespace std;

void reverse(int arr[], int start, int end) {
    while (start < end) {
        swap(arr[start], arr[end]);
        start++;
        end--;
    }
}

void pancakeSort(int arr[], int n) {
    for (int currSize = n; currSize > 1; --currSize) {
        int maxIndex = 0;
        for (int i = 0; i < currSize; ++i) {
            if (arr[i] > arr[maxIndex]) {
                maxIndex = i;
            }
        }

        if (maxIndex != currSize - 1) {
            reverse(arr, 0, maxIndex);
            reverse(arr, 0, currSize - 1);
        }
    }
}

void printArray(int arr[], int n) {
    for (int i = 0; i < n; ++i) {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int main() {
    int arr[] = {3, 6, 2, 7, 5, 8, 1, 4};
    int n = sizeof(arr) / sizeof(arr[0]);

    pancakeSort(arr, n);

    printArray(arr, n);

    return 0;
}

File: PHP/pancake_sort.php
<?php

function flip(&$arr, $i) {
    $start = 0;
    while ($start < $i) {
        $temp = $arr[$start];
        $arr[$start] = $arr[$i];
        $arr[$i] = $temp;
        $start++;
        $i--;
    }
}

function findMaxIndex($arr, $n) {
    $maxIndex = 0;
    for ($i = 1; $i < $n; $i++) {
        if ($arr[$i] > $arr[$maxIndex]) {
            $maxIndex = $i;
        }
    }
    return $maxIndex;
}

function pancakeSort(&$arr) {
    $n = count($arr);
    for ($currSize = $n; $currSize > 1; $currSize--) {
        $maxIndex = findMaxIndex($arr, $currSize);

        if ($maxIndex != $currSize - 1) {
            flip($arr, $maxIndex);
            flip($arr, $currSize - 1);
        }
    }
}

$arr = [3, 6, 1, 8, 4, 5, -9, 0, -3];

pancakeSort($arr);

echo "Sorted Array: " . implode(", ", $arr) . "\n";
?>

